visions/accessors.md

```diff
@@ -0,0 +1,348 @@
+|**Copying**   |get    |set   |✗      |
+|**Borrowing** |borrow |✗     |mutate |
+|**Yielding**  |read   |✗     |modify |
+
+Some of the above exist in the current Swift language, others we expect to be proposed in the near future.  Here is a slightly more detailed explanation of each one:
+
+* `get` and `set` - These are the original standard accessors, which compile into methods that return or accept a value of the indicated type.
+* `borrow` and `mutate` - These borrowing accessors are not yet implemented, but we expect to propose them for Swift Evolution at some point in the future.  These provide access to the value without copying by relying on Swift’s “borrow” implementation which avoids pointer overhead for simple types.  In particular, these can be used for non-copyable and non-escapable values.  (In various discussions, the proposed `mutate` accessor has been called `inout`.)
+* `read` and `modify` -  These exist today as experimental `_read` and `_modify` implementations and we expect to propose a revised final form without the underscored names in the near future. They compile into coroutines that “yield” the value.  In effect, each such accessor becomes two functions:  The “top half of the coroutine” creates a value if necessary and then provides the caller with a pointer to this value.  The “bottom half of the coroutine” cleans up the value.  In effect, callers invoke the top half to obtain a pointer to the value, use the pointer to read and/or modify the value, then call the bottom half to give the property a chance to clean itself up.
+
+In the process of developing the above, a number of other approaches have been explored.  Some of the following are currently supported in the compiler, but none are expected to go through Swift Evolution
+
+* `unsafeAddress` and `unsafeMutableAddress` - These are implemented but are not expected to ever go through Swift Evolution.  They compile into methods that return a pointer to the value in question.  This pointer value is explicit in the property implementation but is not visible in the Swift source code of the client.  This does not require copying the value, but does require that the value already be present somewhere in memory.  This was an early form of the idea behind `borrow` and `mutate` but based on unsafe constructs.
+* `unsafeRawAddress` and `unsafeRawMutableAddress` - These are not yet implemented and may not ever go through Swift Evolution.  These are similar to the above but allow the property implementation to work directly with an untyped “raw” pointer while the caller sees an access to typed data in memory.  This document will refer to these and the previous two collectively as `unsafe*Address`.
+* `_read` and `_modify` - These were the early experimental forms of `read` and `modify`.  They will likely continue to be supported in order to avoid breaking existing code, but users should migrate to `read` and `modify` once they are finalized.
```

Comment:
Will it be an ABI break to move from the underscored versions to the finalized versions?

Reply:
We do plan to provide a way to make this migration while preserving ABI (for people who need to do so).